ENH, DOC: Improve foreign object interoperability

miccoli opened the pull request:
As discussed in GH-28024, the logic with which functions in `numpy/_core/fromnumeric.py` navigate methods in foreign objects is not clearly documented.

The major problem arises for object that *do not* implement the [`__array_function__`](https://numpy.org/devdocs/user/basics.interoperability.html#the-array-function-protocol)  protocol: the fact that say `np.mean(obj)` will call `obj.mean(axis=None, dtype=None, out=None)` even *before* trying `obj.__array__` is not documented.

In this PR:
- [x] add checks in numpy code to avoid calling non callable attributes
- [ ] add note in numpy docs to warn developers about order in which methods are tested if `__array_function__` not present
- [ ] gently nudge developers to adopt newer `__array_function__` protocol

miccoli committed:
TST: check that attributes in classes do not mask numpy functions

This test suite demonstrates that functions in
numpy/_core/fromnumerics.py wrongly assume that if an object has an
attribute with the same name of the function it is a callable method.

All test will fail with "TypeError: 'NoneType' object is not callable",
see GH28024.

A subsequent commit will fix these tests.

miccoli committed:
ENH: functions in _core/fromnumeric do not fail on non callable attrs

Add checks to avoid that functions in numpy/_core/fromnumeric.py fail
with TypeError when called on objects with a non-callable attribute with
the same name as the function itself.
See GH28024.

mhvk commented on numpy/_core/fromnumeric.py:
```diff
@@ -49,8 +64,10 @@ def _wrapit(obj, method, *args, **kwds):
 def _wrapfunc(obj, method, *args, **kwds):
-    bound = getattr(obj, method, None)
-    if bound is None:
+    try:
```
Instead of adding an indirection and therefore slowing down the common case where the method is defined correctly, how about going more directly to the goal, and just checking `TypeError` more consistently as one tries calling?  Here, it would seem the code already does the right thing without any changes`, but in other branches, it needs a `try/except` around the trial to call it.

miccoli replied:
My choice was to keep the two logical steps separate:
- first retrive a callable method
- then try if the interface is correct

This is for readability and clarity. I added also the `_get_method` method function to clarly signal that the first step is repeated in different portions of the code.

For the efficiency, please see comment below.

mhvk commented on numpy/_core/fromnumeric.py:
```diff
@@ -72,8 +89,8 @@ def _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs):
 
     if type(obj) is not mu.ndarray:
         try:
```
So, here, I'd tend to write it more like `_wrapfunc` - indeed, perhaps that can be used in some way?

miccoli replied:
It was not my task to completely refactor the fromnumeric code, just to fix the failures of the tests of 86f6764

mhvk reviewed (commented):
This works, but I somewhat dislike that it slows down the most common case, where it does call the attribute correctly. See in-line comments for a possible solution.

miccoli commented:
> This works, but I somewhat dislike that it slows down the most common case, where it does call the attribute correctly. See in-line comments for a possible solution.

These are python functions, not ufuncs, so I think that this concern is misplaced: I tried to favor clarity of intent over speed, while keeping the changes to a minimum.

If necessary I can setup a proper benchmark, however just for curiosity I wrote this simple `timeit` script:

<details>

```python
import timeit

import numpy as np

p0 = """
import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)
"""

bench = ["np.sum(a)", "a.sum()", "np.cumsum(a)", "a.cumsum()"]
n = 100_000
r = 10
scale = 1e6


print(np.version.version)
print(p0)
print(f"n = {n:_d}, r = {r:d}")

for s in bench:
    print(f"{s:12s}", end="")
    t = timeit.Timer(stmt=s, setup=p0)
    t1 = np.array(t.repeat(repeat=r, number=n)) / n * scale
    print(f"  min = {t1.min():.3f} µs, mean = ({t1.mean():.3f} ± {t1.std():.3f}) µs ")
```

</details>

The results on my (old) laptop are:

#### baseline

```
2.3.0.dev0+git20250102.12364e3

import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)

n = 100_000, r = 10
np.sum(a)     min = 2.847 µs, mean = (2.919 ± 0.042) µs
a.sum()       min = 1.483 µs, mean = (1.533 ± 0.044) µs
np.cumsum(a)  min = 4.017 µs, mean = (4.068 ± 0.050) µs
a.cumsum()    min = 3.021 µs, mean = (3.389 ± 0.973) µs
```

#### after this PR

```
2.3.0.dev0+git20250103.b9a8d9e

import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)

n = 100_000, r = 10
np.sum(a)     min = 3.153 µs, mean = (3.328 ± 0.098) µs
a.sum()       min = 1.743 µs, mean = (1.778 ± 0.030) µs
np.cumsum(a)  min = 4.120 µs, mean = (4.200 ± 0.057) µs
a.cumsum()    min = 3.096 µs, mean = (3.134 ± 0.031) µs
```

My comment here is that this PR for sure add a small overhead, but this is so small (< 0.5 µs per call) that it is very hard to measure it properly, at least on a laptop.

Here is a second run of the very same benchmark, after the system awaked from sleep: you can see that now the code after the PR is even faster than the baseline. This clearly shows that on a multi-user system, the operating conditions have a big influence on the figures.
<details>

```
2.3.0.dev0+git20250103.b9a8d9e

import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)

n = 100_000, r = 10
np.sum(a)     min = 2.700 µs, mean = (2.751 ± 0.038) µs
a.sum()       min = 1.489 µs, mean = (1.538 ± 0.029) µs
np.cumsum(a)  min = 3.997 µs, mean = (4.106 ± 0.073) µs
a.cumsum()    min = 2.958 µs, mean = (3.010 ± 0.032) µs
```
</details>


### final considerations

Properly benchmarking this changes would be very a complex task (meaningful bechmarking is an art in itself...) but I do not think that it makes much sense: if in a given “numpy intensive” program the bottleneck are the overheads due to python function calls, then probably the program should be refactored, instead of over-optimizing the numpy code-base.

mhvk commented:
But why increase the overhead at all if it is not needed? This is code that is used a *lot*, so speed does matter. (Also, you use fairly large arrays; for small ones it would be worse.)

miccoli commented:
> But why increase the overhead at all if it is not needed? This is code that is used a _lot_, so speed does matter. (Also, you use fairly large arrays; for small ones it would be worse.)

My point is that nobody will notice. Please have a look at the above timings: actually only `np.cumsum(a)` goes through the new branches (via `_wrapfunc`). All the other function calls should not be touched by the changes. The small timing differences are close to noise.

mhvk commented:
Please believe me: people *will* notice an extra 0.3 us -- that's a 17% increase for `a.sum()` (for a relatively large array; the fractional increase will be worse for smaller ones).  There's simply no reason to pay that price for a work-around for a problem that affects very few people, especially if for those performance doesn't matter at all since an error is going to get raised anyway.

miccoli commented:
> Please believe me: people _will_ notice an extra 0.3 us -- that's a 17% increase for `a.sum()` (for a relatively large array; the fractional increase will be worse for smaller ones).

`a.sum()` is not affected by the changes, it follows a different path, not affected by this PR. I'm timing it because `np.sum(a)` will dispatch to `a.sum()`, to have a reference.

Other timings, with `n = 1_000_000` in which the differences become smaller:
<details>

```
2.3.0.dev0+git20250102.12364e3

# import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)

n = 1_000_000, r = 10
np.sum(a)     min = 2.956 µs, mean = (3.054 ± 0.095) µs
a.sum()       min = 1.569 µs, mean = (1.663 ± 0.151) µs
np.cumsum(a)  min = 4.106 µs, mean = (4.230 ± 0.148) µs
a.cumsum()    min = 3.079 µs, mean = (3.135 ± 0.103) µs
```

```
2.3.0.dev0+git20250103.b9a8d9e

# import numpy as np
a = np.arange(1_000, dtype=np.float64).reshape(20, -1)

n = 1_000_000, r = 10
np.sum(a)     min = 3.053 µs, mean = (3.188 ± 0.123) µs
a.sum()       min = 1.584 µs, mean = (1.611 ± 0.025) µs
np.cumsum(a)  min = 4.134 µs, mean = (4.237 ± 0.117) µs
a.cumsum()    min = 3.093 µs, mean = (3.236 ± 0.157) µs
```

`timeit` seems not to be adequate here to catch the overhead... As said a proper benchmark, encompassing a variety of matrix sizes and different functions should be set up. I'm confident that the measured overhead should be much smaller than the 0.5 µs that I mentioned above as a very loose upper limit ...

</details>

Sorry but I remain of my opinion: you will notice the overhead only if you call `np.cumsum(a)` in a tight loop with smallish `a` (and for me 20×50 is smallish), but in such cases a code refactor for vectorizing the loop, or rewriting in a compiled language would be the optimal solution.

BTW thanks for the discussion.

miccoli committed:
Slightly reduce footprint of changes

Avoid calling _wrapit from an error handler, to be more consistent with
original logic.

seberg commented on numpy/_core/fromnumeric.py:
```diff
@@ -72,8 +88,8 @@ def _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs):
-            reduction = getattr(obj, method)
-        except AttributeError:
+            reduction = _get_method(obj, method)
+        except (AttributeError, TypeError):
```
I wonder if we should discuss briefly what to do in the non-callable path (fall back or just raise something more helpful even?).

But with the exception of `put`, where I am not sure any change is even needed, I think the: `func = getattr(obj, "attr", None); if callable(func): ...` version should just be adopted here also.  It is much simpler to read, avoids the try/except and should have minimal overhead when the method exists (much less than even a single custom Python function call for sure).

miccoli replied:
The original logic is that `np.fun(a, ...)` should first check if it is possible to dispatch to `a.fun(...)` and then fall back to `_wrapit` (which tries to convert the object to an array). From this point of view, it seems to me correct to treat missing/non callable attributes in the same path.

What is not clear is if the call to `a.fun(...)` should *always* be included in a `try`/`except` block: this is done in `_wrapfunc` to recover from calling methods that do not fully implement the numpy API. However In other points, this guard is not present.

As what regards the second comment, in my first take I tried to DRY up and always use `_get_method`. However after be89241 it si probably more legible and clear to always use the `func = getattr(obj, "attr", None); if callable(func): ...` approach.

miccoli replied:
Commit 0c319d9 makes liberal use of `if callable(fun)` instead of wrapping in `try:`/`except:`.

In fact the code seems to me more readble after 0c319d9.

miccoli committed:
Look before you leap: always check for callable method

Be more explicit on when we want to call obj.method, even if this
results in a "look before you leap" approach.
More liberal use of if callable(method) results in clearer code.
